### PR TITLE
Reduce PersistentTimestampService.getFreshTimestamps contention

### PR DESCRIPTION
**Pull request by schlosna**
**Goals (and why)**: Reduce contention on `PersistentTimestampService.getFreshTimestamps`

**How**: Minimize critical sections of `getFreshTimestamps` code paths

**Concerns**: These code paths are critical to the correctness of timestamp service, are hammered from many concurrent threads, and become bottlenecks under timestamp request load.

**Priority**: this week

**Entrypoint (where should we start reviewing?)**: PersistentTimestampService.getFreshTimestamps

**Context**
While profiling the timestamp service under heavy load for timelock server scale testing, there was significant contention on PersistentTimestampService.getFreshTimestamps.

The contention and critical sections have been reduced to the minimal required to preserve correctness, and relevant classes have been properly annotated with javax.annotation.concurrent.* annotations indicating thread-safety or required guard conditions.

---
This change is [Reviewable](https://reviewable.io/reviews/palantir/atlasdb/1618)

### CODE REVIEW

File: timestamp-impl/src/main/java/com/palantir/timestamp/AvailableTimestamps.java [hunk 1]
```diff
@@ -65,25 +90,29 @@ public synchronized void fastForwardTo(long newMinimum) {
         upperLimit.increaseToAtLeast(newMinimum + ALLOCATION_BUFFER_SIZE);
     }
 
-    private long lastHandedOut() {
+    private synchronized long lastHandedOut() {
```
Comment (schlosna) [inline timestamp-impl/src/main/java/com/palantir/timestamp/AvailableTimestamps.java]: note to reviewers: technically this was already safe, but want `synchronized` for documentation and avoids having to litter with `assert Thread.holdsLock(this);` that bloats bytecode more.

File: timestamp-impl/src/main/java/com/palantir/timestamp/AvailableTimestamps.java [hunk 2]
```diff
@@ -65,25 +90,29 @@ public synchronized void fastForwardTo(long newMinimum) {
...
+    private synchronized TimestampRange handOutTimestamp(long targetTimestamp) {
+        TimestampRange rangeToHandOut = getRangeToHandOut(targetTimestamp);
         allocateEnoughTimestampsToHandOut(targetTimestamp);
         lastReturnedTimestamp.increaseToAtLeast(targetTimestamp);
 
         return rangeToHandOut;
     }
 
-    private void allocateEnoughTimestampsToHandOut(long timestamp) {
+    private synchronized void allocateEnoughTimestampsToHandOut(long timestamp) {
```
Comment (schlosna) [inline timestamp-impl/src/main/java/com/palantir/timestamp/AvailableTimestamps.java]: note to reviewers: technically this was already safe, but want `synchronized` for documentation and avoids having to litter with `assert Thread.holdsLock(this);` that bloats bytecode more.

File: timestamp-impl/src/main/java/com/palantir/timestamp/LastReturnedTimestamp.java [hunk 3]
```diff
@@ -23,10 +26,13 @@ public LastReturnedTimestamp(long timestamp) {
...
     public synchronized void increaseToAtLeast(long newTimestamp) {
-        this.timestamp = Math.max(this.timestamp, newTimestamp);
+        long current = this.timestamp;
+        if (newTimestamp > current) {
+            this.timestamp = newTimestamp;
+        }
     }
 
-    public long get() {
+    public synchronized long get() {
```
Comment (schlosna) [inline timestamp-impl/src/main/java/com/palantir/timestamp/LastReturnedTimestamp.java]: note to reviewers: previously there was a potential race condition as we were doing writes under `synchronized(this)` but reads were not -- this resolves that potential misuse.

File: timestamp-impl/src/main/java/com/palantir/timestamp/LastReturnedTimestamp.java [hunk 4]
```diff
@@ -23,10 +26,13 @@ public LastReturnedTimestamp(long timestamp) {
     }
 
     public synchronized void increaseToAtLeast(long newTimestamp) {
-        this.timestamp = Math.max(this.timestamp, newTimestamp);
+        long current = this.timestamp;
+        if (newTimestamp > current) {
+            this.timestamp = newTimestamp;
```
Comment (schlosna) [inline timestamp-impl/src/main/java/com/palantir/timestamp/LastReturnedTimestamp.java]: note to reviewers: only write field when actually needed

**Comment by schlosna** [comment]
The following shows contention from a 3 minute snapshot of TimelockServer

```
Stack Trace                                                                                            Count    Duration
com.palantir.timestamp.PersistentTimestampService.getFreshTimestamps(int)                              206,768  5h 56min 29s 20ms
   com.palantir.timestamp.PersistentTimestampService.getFreshTimestamp()                               206,768  5h 56min 29s 20ms
      com.palantir.atlasdb.timelock.paxos.DelegatingManagedTimestampService.getFreshTimestamp()        206,768  5h 56min 29s 20ms
         com.palantir.leader.proxy.AwaitingLeadershipProxy.handleInvocation(Object, Method, Object[])  206,768  5h 56min 29s 20ms
```

File: timestamp-impl/src/main/java/com/palantir/timestamp/PersistentTimestampService.java [hunk 5]
```diff
@@ -37,19 +38,20 @@ public PersistentTimestampService(AvailableTimestamps availableTimestamps, Execu
...
 
     public static PersistentTimestampService create(TimestampBoundStore tbs) {
         PersistentUpperLimit upperLimit = new PersistentUpperLimit(tbs);
         LastReturnedTimestamp lastReturned = new LastReturnedTimestamp(upperLimit.get());
         AvailableTimestamps availableTimestamps = new AvailableTimestamps(lastReturned, upperLimit);
-        ExecutorService executor = Tracers.wrap(
-                PTExecutors.newSingleThreadExecutor(
-                        PTExecutors.newThreadFactory("Timestamp allocator", Thread.NORM_PRIORITY, true)));
+        ExecutorService executor = PTExecutors.newSingleThreadExecutor(
+                PTExecutors.newThreadFactory("Timestamp allocator", Thread.NORM_PRIORITY, true));
```
Comment (schlosna) [inline timestamp-impl/src/main/java/com/palantir/timestamp/PersistentTimestampService.java]: note to reviewers: removing tracing for now as it was allocating a bunch and is likely unnecessary.

File: timestamp-impl/src/main/java/com/palantir/timestamp/PersistentTimestampService.java [hunk 6]
```diff
@@ -60,7 +62,7 @@ public long getFreshTimestamp() {
     }
 
     @Override
-    public synchronized TimestampRange getFreshTimestamps(int numTimestampsRequested) {
+    public TimestampRange getFreshTimestamps(int numTimestampsRequested) {
         int numTimestampsToHandOut = cleanUpTimestampRequest(numTimestampsRequested);
         TimestampRange handedOut = availableTimestamps.handOut(numTimestampsToHandOut);
         asynchronouslyRefreshBuffer();
```
Comment (schlosna) [inline timestamp-impl/src/main/java/com/palantir/timestamp/PersistentTimestampService.java]: note to reviewers: this was the major source of contention via the `synchronized` on the method. There is now no need for method level locking as `cleanUpTimestampRequest` is a static that checks argument, `availableTimestamps.handOut` does its own locking, and `asynchronouslyRefreshBuffer()` now handles ensuring only a single refresh task is in-flight at a time.

File: timestamp-impl/src/main/java/com/palantir/timestamp/PersistentTimestampService.java [hunk 7]
```diff
@@ -75,15 +77,25 @@ public void fastForwardTimestamp(long currentTimestamp) {
...
     }
 
-    private int cleanUpTimestampRequest(int numTimestampsRequested) {
-        Preconditions.checkArgument(numTimestampsRequested > 0,
-                "Number of timestamps requested must be greater than zero, was %s",
-                numTimestampsRequested);
+    private static int cleanUpTimestampRequest(int numTimestampsRequested) {
+        if (numTimestampsRequested <= 0) {
+            throw new IllegalArgumentException(String.format(
+                    "Number of timestamps requested must be greater than zero, was %s", numTimestampsRequested));
```
Comment (schlosna) [inline timestamp-impl/src/main/java/com/palantir/timestamp/PersistentTimestampService.java]: note to reviewers: goal here is avoid allocations on happy path

File: timestamp-impl/src/main/java/com/palantir/timestamp/TimestampAllocationFailures.java [hunk 8]
```diff
@@ -68,14 +73,11 @@ private void logNewFailure(Throwable newFailure) {
     }
 
     private boolean isSameAsPreviousFailure(Throwable newFailure) {
-        if(previousAllocationFailure == null) {
-            return false;
-        }
-
-        return newFailure.getClass().equals(previousAllocationFailure.getClass());
+        Throwable failure = this.previousAllocationFailure;
+        return failure != null && failure.getClass().equals(newFailure.getClass());
```
Comment (schlosna) [inline timestamp-impl/src/main/java/com/palantir/timestamp/TimestampAllocationFailures.java]: note to reviewers: now use only single volatile read

File: timestamp-impl/src/main/java/com/palantir/timestamp/TimestampAllocationFailures.java [hunk 9]
```diff
@@ -15,50 +15,55 @@
...
-   }
+    }
 
-    public synchronized void verifyWeShouldTryToAllocateMoreTimestamps() {
-        if(previousAllocationFailure instanceof MultipleRunningTimestampServiceError) {
-            throw wrapMultipleRunningTImestampServiceError(previousAllocationFailure);
+    public void verifyWeShouldTryToAllocateMoreTimestamps() {
+        Throwable failure = this.previousAllocationFailure;
+        if (failure instanceof MultipleRunningTimestampServiceError) {
+            throw wrapMultipleRunningTimestampServiceError(failure);
```
Comment (schlosna) [inline timestamp-impl/src/main/java/com/palantir/timestamp/TimestampAllocationFailures.java]: note to reviewers: now use only single volatile read

**Comment by jeremyk-91** [comment]
@schlosna I'll review this more carefully tomorrow morning; offhand, the logic for splitting the critical sections to each component makes sense.

I'll also point our Jepsen job runner to this branch to kick off a test.

File: timestamp-impl/src/main/java/com/palantir/timestamp/AvailableTimestamps.java [hunk 10]
```diff
@@ -34,27 +38,48 @@ public AvailableTimestamps(LastReturnedTimestamp lastReturnedTimestamp, Persiste
...
 
-        long targetTimestamp = lastHandedOut() + numberToHandOut;
-        DebugLogger.logger.trace("Handing out {} timestamps, taking us to {}.", numberToHandOut, targetTimestamp);
-        return handOutTimestamp(targetTimestamp);
+    public TimestampRange handOut(long numberToHandOut) {
+        checkValidTimestampRangeRequest(numberToHandOut);
+        synchronized (this) {
+            long targetTimestamp = lastHandedOut() + numberToHandOut;
+            DebugLogger.logger.trace("Handing out {} timestamps, taking us to {}.", numberToHandOut, targetTimestamp);
+            return handOutTimestamp(targetTimestamp);
```
Comment (jnelson15) [inline timestamp-impl/src/main/java/com/palantir/timestamp/AvailableTimestamps.java]: May not matter relative to the other more expensive operations, but could pull the logging out of the synchronized block if we're trying to keep this as minimized as possible.

Comment (schlosna) [inline timestamp-impl/src/main/java/com/palantir/timestamp/AvailableTimestamps.java]: Yeah, I might pull the logging out

File: timestamp-impl/src/main/java/com/palantir/timestamp/AvailableTimestamps.java [hunk 11]
```diff
@@ -34,27 +38,48 @@ public AvailableTimestamps(LastReturnedTimestamp lastReturnedTimestamp, Persiste
...
+        }
 
-        if (buffer < MINIMUM_BUFFER || !upperLimit.hasIncreasedWithin(1, MINUTES)) {
-            DebugLogger.logger.trace(
-                    "refreshBuffer: refreshing and allocating timestamps. Buffer {}, Current upper limit {}.",
-                    buffer,
-                    currentUpperLimit);
-            allocateEnoughTimestampsToHandOut(lastHandedOut() + ALLOCATION_BUFFER_SIZE);
+        if (DebugLogger.logger.isTraceEnabled()) {
+            logRefresh(needsRefresh, currentUpperLimit, buffer);
```
Comment (jeremyk-91) [inline timestamp-impl/src/main/java/com/palantir/timestamp/AvailableTimestamps.java]: Right. Yep, I see.

File: timestamp-impl/src/main/java/com/palantir/timestamp/LastReturnedTimestamp.java (hunk 4 above)
Comment (jeremyk-91) [inline timestamp-impl/src/main/java/com/palantir/timestamp/LastReturnedTimestamp.java]: OK. I don't suppose there are significant gains to do away with `current` altogether?

Comment (schlosna) [inline timestamp-impl/src/main/java/com/palantir/timestamp/LastReturnedTimestamp.java]: will inline

**Pull request merged**